osu.Game/Overlays/DialogOverlay.cs

```diff
@@ -49,18 +49,24 @@ public void Push(PopupDialog dialog)
         {
             if (dialog == CurrentDialog || dialog.State.Value != Visibility.Visible) return;
 
-            // if any existing dialog is being displayed, dismiss it before showing a new one.
-            CurrentDialog?.Hide();
-
+            var lastDialog = CurrentDialog;
             CurrentDialog = dialog;
-            CurrentDialog.State.ValueChanged += state => onDialogOnStateChanged(dialog, state.NewValue);
 
-            dialogContainer.Add(CurrentDialog);
+            Scheduler.Add(() =>
+            {
+                // if any existing dialog is being displayed, dismiss it before showing a new one.
+                lastDialog?.Hide();
+
+                CurrentDialog = dialog;
```

Comment:
`CurrentDialog` is set to `dialog` twice in this method: once immediately, and once inside the schedule. Is this intended? Scares me a bit, especially what with the `dialog == CurrentDialog` guard at the start of this method that may not work correctly now...

Also if the outer write is the only correct one, then I'd expect to see an "ensure we haven't been replaced already" flow here, as in something along the lines of

```csharp
if (CurrentDialog == dialog)
    // set up value changed callbacks, add to container, show
```